### PR TITLE
Fix target group leaking issue

### PR DESCRIPTION
**Pull request by zijun726911**
Trying to solve this Target Groups Leaked  issue: https://github.com/aws/aws-application-networking-k8s/issues/119

## Changes:

- Changed model_build_targetgroup and model_build_service, make sure target group deletion request `model` triggered by httproute deletion could pass to target_group_synthesizer

- Changed lattice resource creation and deletion order. move delete latticeSerivce and delete Targetgroup logic in the in the corresponding PostSynthesize()
    - __New LatticeResource creation order when receive httproute creation request:__ TargetGroup -> Targets -> LatticeService -> Listener -> Rule
   
    - __New LatticeResource deletion order when receive httproute deletion request:__ LatticeService (vpc lattice backend will asynchronous delete its listeners and rules ) ->  TargetGroup(target_group_manager.go layer would check and deregister targets for this target group)
    
- Split the SynthesizeTriggeredTargetGroup into two functions: SynthesizeTriggeredTargetGroupCreation (invoked in Synthesize() ) , SynthesizeTriggeredTargetGroupDeletion (invoked in PostSynthesize() )
-  Add target group clean up verification logic in the e2etest framework ToBeClean()
- Add a new HttpRoute mutation e2etest case
-  Add some code comments either related to my code change or general comments to improve readability

---

## Test:
- `make e2etest` pass, all 4 tests could pass. Full log : https://gist.github.com/zijun726911/abcf41ba1aa1fdadab364128db2e2796

- gateway related all previsous passed conformance test could pass
-  Per #119 , do `kubectl apply -f myapp.yaml` and then do  the `kubectl delete -f myapp.yaml`, all lattice resource could be deleted, no any lattice resource leaking, where myappl.yaml contains:
     - Gateway
     - HTTPRoute
     - Service
     - Deployment
- do `kubectl apply -f myapp.yaml`  and then do the `kubectl delete httproute myhttproute`, all lattice resource could be deleted, no any lattice resource leaking

---

## New LatticeResource creation and deletion order from the `latticeServiceStackDeployer` perspective

From the `latticeServiceStackDeployer` perspective, when receive a httproute creation request, the LatticeResource creation order is :

1. TargetGroupSynthesizer.Synthesize()  (handle tg creation request, create tg )
2. TargetsSynthesizer.Synthesize()  (handle targets 'reconcile' request, (reconcile means invoke vpc lattice apis to create or delete Targets to make it change to the _desigired_ num of targets ) )
3. ServiceSynthesizer.Synthesize()  (handle service creation request, create latticeService)
4.  ListenerSynthesizer.Synthesize()  (handle listener creation request, create listener )
5.  RuleSynthesizer.Synthesize()  (handle rules creation request, create rules )
6. ServiceSynthesizer.PostSynthesize() (skip service creation request)
7. TargetGroupSynthesizer.PostSynthesize()  (skip tg cretion request )

From the `latticeServiceStackDeployer` perspective, when receive a httproute deletion request,  the LatticeResource deletion order is :

1. TargetGroupSynthesizer.Synthesize()  (skip tg deletion request)
2. TargetsSynthesizer.Synthesize()  ([do nothing, as  model_build_lattice_service layer didn't build targets model](https://github.com/aws/aws-application-networking-k8s/blob/1af736bd18b6f8ea7eea6602ecd5f1c60377c2ed/pkg/gateway/model_build_lattice_service.go#L88))
3. ServiceSynthesizer.Synthesize()  (skip service deletion request)
4.  ListenerSynthesizer.Synthesize()  ([do nothing as  model_build_lattice_service layer didn't build listener model ](https://github.com/aws/aws-application-networking-k8s/blob/1af736bd18b6f8ea7eea6602ecd5f1c60377c2ed/pkg/gateway/model_build_lattice_service.go#L88))
5.  RuleSynthesizer.Synthesize()  ([do nothing as  model_build_lattice_service layer didn't build rule model](https://github.com/aws/aws-application-networking-k8s/blob/1af736bd18b6f8ea7eea6602ecd5f1c60377c2ed/pkg/gateway/model_build_lattice_service.go#L88))
6. ServiceSynthesizer.PostSynthesize() (__handle LatticeService deletion request, delete the LatticeService, and then, the VPC Lattice backend it will asynchronous delete the listener and rules and disassociate service-targetgroup relation__)
7. TargetGroupSynthesizer.PostSynthesize()  (__handle tg deletion request, check if the tg have still unused status targets, if yes, deregister them, if the tg have still have other status targets, return a retry, if all targets clear, delete the targetgroup__)

### CODE REVIEW

**Comment by liwenwu-amazon** [comment]
QQ, 
* any reason you moved lots of logic to PostSynthersize() routine?  Also, have you added unit test(s) to cover all logics in PostSynthersize()?
* is this build passing e2e and no regression on conformance testing?

**Comment by zijun726911** [comment]
> QQ,
> 
> * any reason you moved lots of logic to PostSynthersize() routine?  Also, have you added unit test(s) to cover all logics in PostSynthersize()?
> * is this build passing e2e and no regression on conformance testing?

Currently 3 e2etest cases could pass. 
The purpose of moving latticeResource deletion logic is make latticeResource creation and deletion have reversed order.

I am still working on the code change. I will supplement all these detail in the code comments and PR description

**Comment by zijun726911** [comment]
> 

I added something like these in the current unit tests:
```
if tt.httpRoute.DeletionTimestamp.IsZero() {
			err = synthesizer.Synthesize(ctx)
		} else {
			err = synthesizer.PostSynthesize(ctx)
		}
```

**Comment by liwenwu-amazon** [comment]
> 

since you reverse the order during delete, i would expect of the unit tests of these functions would change.

File: pkg/gateway/model_build_lattice_service.go [hunk 1]
```diff
@@ -76,13 +76,6 @@ func (t *latticeServiceModelBuildTask) buildModel(ctx context.Context) error {
 		return err
 	}
 
-	if !t.httpRoute.DeletionTimestamp.IsZero() {
-		// in case of deleting HTTPRoute, we will let reconcile logic to delete
-		// stated target group(s) at next reconcile interval
-		glog.V(2).Infof("latticeServiceModuleBuildTask: for HTTPRouteDelete, reconcile tagetgroups/targets at reconcile interval")
-		return nil
-	}
-
```
Comment (zijun726911) [inline pkg/gateway/model_build_lattice_service.go:85]: Removed this if condition, need to go ahead to build the TargetGroup model to make target_group_synthesizer could receive this  target group deletion request

File: pkg/gateway/model_build_targetgroup.go [hunk 2]
```diff
@@ -248,15 +248,19 @@ func (t *latticeServiceModelBuildTask) buildTargetGroup(ctx context.Context, cli
...
 				// to add
-				t.Datastore.SetTargetGroupByBackendRef(tgName, t.httpRoute.Name, false, true)
+				t.Datastore.SetTargetGroupByBackendRef(tgName, t.httpRoute.Name, tgSpec.Config.IsServiceImport, true)
 			} else {
 				// to delete
-				t.Datastore.SetTargetGroupByBackendRef(tgName, t.httpRoute.Name, false, false)
+				t.Datastore.SetTargetGroupByBackendRef(tgName, t.httpRoute.Name, tgSpec.Config.IsServiceImport, false)
+				dsTG, _ := t.Datastore.GetTargetGroup(tgName, t.httpRoute.Name, tgSpec.Config.IsServiceImport)
+				tgSpec.IsDeleted = true
+				tgSpec.LatticeID = dsTG.ID
```
Comment (zijun726911) [inline pkg/gateway/model_build_targetgroup.go:263]: target_group_manger need the `LatticeID` to invoke vpcLattice.deleteTargetgroup Api to handle the target group deletion request 

Change false to tgSpec.Config.IsServiceImport in order to build tg model for both backendref service and ServiceImport, for  ServiceImport, target_group_synthesizer.go will ignore to delete this tg.

File: pkg/gateway/model_build_targetgroup.go [hunk 3]
```diff
@@ -320,20 +335,16 @@ func (t *latticeServiceModelBuildTask) buildHTTPTargetGroupSpec(ctx context.Cont
...
 	tgName := latticestore.TargetGroupName(string(httpBackendRef.Name), namespace)
 
-	var isDeleted bool
-
-	if t.httpRoute.DeletionTimestamp.IsZero() {
-		isDeleted = false
-	} else {
-		isDeleted = true
-	}
-
```
Comment (zijun726911) [inline pkg/gateway/model_build_targetgroup.go:336]: Moved this bock to the begining of this function

```
var isDeleted bool

	if t.httpRoute.DeletionTimestamp.IsZero() {
		isDeleted = false
	} else {
		isDeleted = true
	}
```

File: pkg/gateway/model_build_targetgroup.go [hunk 4]
```diff
@@ -320,20 +335,16 @@ func (t *latticeServiceModelBuildTask) buildHTTPTargetGroupSpec(ctx context.Cont
 		svc := &corev1.Service{}
 		if err := t.Client.Get(ctx, serviceNamespaceName, svc); err != nil {
 			glog.V(6).Infof("Error finding backend service %v error :%v \n", serviceNamespaceName, err)
-			return latticemodel.TargetGroupSpec{}, err
+			if !isDeleted {
+				//Return error for creation request only,
+				//For k8sService deletion request, we should go ahead to build TargetGroupSpec model
+				return latticemodel.TargetGroupSpec{}, err
```
Comment (zijun726911) [inline pkg/gateway/model_build_targetgroup.go:341]: return error for creation request only,
For k8sService deletion request, we should go ahead to build TargetGroupSpec model

File: pkg/gateway/model_build_targetgroup.go [hunk 5]
```diff
@@ -302,7 +313,11 @@ func (t *latticeServiceModelBuildTask) buildHTTPTargetGroupSpec(ctx context.Cont
 
 		} else {
 			glog.V(6).Infof("buildHTTPTargetGroupSpec, using service Import %v, err :%v\n", namespaceName, err)
-			return latticemodel.TargetGroupSpec{}, err
+			if !isDeleted {
+				//Return error for creation request only.
+				//For ServiceImport deletion request, we should go ahead to build TargetGroupSpec model
+				return latticemodel.TargetGroupSpec{}, err
```
Comment (zijun726911) [inline pkg/gateway/model_build_targetgroup.go:319]: Return error for creation request only.
For ServiceImport deletion request, we should go ahead to build TargetGroupSpec model

File: pkg/gateway/model_build_targetgroup_test.go [hunk 6]
```diff
@@ -277,9 +276,8 @@ func Test_TGModelByHTTPRouteBuild(t *testing.T) {
 			name: "Create LatticeService where backend K8S service does NOT exist",
 			httpRoute: &gateway_api.HTTPRoute{
 				ObjectMeta: metav1.ObjectMeta{
```
Comment (zijun726911) [inline pkg/gateway/model_build_targetgroup_test.go:278]: When test httproute creation (wantIsDeleted is false),  HTTPRoute's DeletionTimestamp should be 0

File: pkg/gateway/model_build_targetgroup_test.go [hunk 7]
```diff
@@ -316,9 +314,8 @@ func Test_TGModelByHTTPRouteBuild(t *testing.T) {
 			name: "Create LatticeService where backend mcs serviceimport does NOT exist",
 			httpRoute: &gateway_api.HTTPRoute{
 				ObjectMeta: metav1.ObjectMeta{
-					Name:              "service4",
-					Finalizers:        []string{"gateway.k8s.aws/resources"},
-					DeletionTimestamp: &now,
+					Name:       "service4",
+					Finalizers: []string{"gateway.k8s.aws/resources"},
 				},
```
Comment (zijun726911) [inline pkg/gateway/model_build_targetgroup_test.go:319]: When test httproute creation (wantIsDeleted is false),  HTTPRoute's DeletionTimestamp should be 0

File: test/pkg/test/framework.go [hunk 8]
```diff
@@ -138,17 +138,30 @@ func (env *Framework) ExpectToBeClean(ctx context.Context) {
...
+				Logger(ctx).Infof("Found Tags for tg %v tags: %v", *tg.Name, retrievedTags)
+				tagValue, ok := retrievedTags.Tags[lattice.K8SParentRefTypeKey]
+				if ok && *tagValue == lattice.K8SServiceExportType {
+					//This tg is created by k8s controller, by a ServiceExport,
+					//ServiceExport still have a known targetGroup leak issue,
+					//so we temporarily skip to verify whether ServiceExport created TargetGroup is deleted or not
+					continue
+				}
+				Expect(*tg.Name).To(Not(ContainElements(BeKeyOf(env.TestCasesCreatedServiceNames))))
+			}
```
Comment (zijun726911) [inline test/pkg/test/framework.go:162]: New test framework works for the new controller code, it could correct skip other cluster's tg and serviceExport created tg:


```
   logger.go:130: 2023-06-15T19:49:05.341-0700	INFO	test/framework.go:144	Found TargetGroup: {
          Arn: "arn:aws:vpc-lattice:us-west-2:348739612106:targetgroup/tg-02b3e0afe03a6cf58",
          CreatedAt: 2023-06-16 02:40:29.099 +0000 UTC,
          Id: "tg-02b3e0afe03a6cf58",
          IpAddressType: "IPV4",
          LastUpdatedAt: 2023-06-16 02:40:29.099 +0000 UTC,
          Name: "k8s-glaze-11-qzxpkn4pvo-default",
          Port: 80,
          Protocol: "HTTP",
          ServiceArns: [],
          Status: "ACTIVE",
          Type: "IP",
          VpcIdentifier: "vpc-0c9c66ebcf3b99ae2"
        }, checking it whether it's created by current EKS Cluster
    logger.go:130: 2023-06-15T19:49:05.459-0700	INFO	test/framework.go:153	Found Tags for tg k8s-glaze-11-qzxpkn4pvo-default tags: {
          Tags: {
            K8SServiceName: "glaze-11-qzxpkn4pvo",
            K8SParentRefTypeKey: "K8SServiceExportType",
            K8SServiceNamespace: "default"
          }
        }
    logger.go:130: 2023-06-15T19:49:05.459-0700	INFO	test/framework.go:144	Found TargetGroup: {
          Arn: "arn:aws:vpc-lattice:us-west-2:348739612106:targetgroup/tg-0f8f06ea953ad8903",
          CreatedAt: 2023-06-09 16:26:47.128 +0000 UTC,
          Id: "tg-0f8f06ea953ad8903",
          IpAddressType: "IPV4",
          LastUpdatedAt: 2023-06-09 16:26:47.128 +0000 UTC,
          Name: "k8s-parking-ver1-default",
          Port: 8090,
          Protocol: "HTTP",
          ServiceArns: [],
          Status: "ACTIVE",
          Type: "IP",
          VpcIdentifier: "vpc-0b46728134fe4943f"
        }, checking it whether it's created by current EKS Cluster
• [253.781 seconds]
------------------------------

Ran 3 of 3 Specs in 772.235 seconds
SUCCESS! -- 3 Passed | 0 Failed | 0 Pending | 0 Skipped
--- PASS: TestIntegration (772.66s)
PASS
ok  	github.com/aws/aws-application-networking-k8s/test/suites/integration	773.397s
```
image
image

**Comment by liwenwu-amazon** [comment]
One general comments
* today, Synthesize<OBJECT>() is meant to  sync up from K8S object to Lattice Object (K8S -->Lattice) 
*  SynthesizeSDK<OBJECT>() is meant to sync up from Lattice Object from K8S object (Lattice--> K8S) 
* PostSynthesize<OBJECT>() , things we do after above 2 steps
---
I would recommend,  for HTTPRoute portion, can you have its own deploy method and explicit call out what you are doing.  e.g.
(d * latticeServiceStackDeployer ) deploy() {
     synthesizecreation()
     synthesizedelete()
     synthsizeSDK() 
}

**Comment by zijun726911** [comment]
> One general comments
> 
> * today, Synthesize() is meant to  sync up from K8S object to Lattice Object (K8S -->Lattice)
> * SynthesizeSDK() is meant to sync up from Lattice Object from K8S object (Lattice--> K8S)
> * PostSynthesize() , things we do after above 2 steps
>   
>   
>   I would recommend,  for HTTPRoute portion, can you have its own deploy method and explicit call out what you are doing.  e.g.
>   (d * latticeServiceStackDeployer ) deploy() {
>   synthesizecreation()
>   synthesizedelete()
>   synthsizeSDK()
>   }



Ok, Do you mean for the latticeServiceStackDeployer, it should not use this stack_deployer.go's `deploy()` method, in stead it should use it's own one?

https://github.com/aws/aws-application-networking-k8s/blob/c4f60928fd2aa377dccf4edab0507c4c7da66b1b/pkg/deploy/stack_deployer.go#L50

File: pkg/deploy/stack_deployer.go [hunk 9]
```diff
@@ -93,14 +93,48 @@ func NewLatticeServiceStackDeploy(cloud aws.Cloud, k8sClient client.Client, latt
...
+	ruleSynthesizer := lattice.NewRuleSynthesizer(d.ruleManager, stack, d.latticeDataStore)
+
+	//Handle targetGroups creation request
+	if err := targetGroupSynthesizer.Synthesize(ctx); err != nil {
+		return err
 	}
-	return deploy(ctx, stack, synthesizers)
+
+	//Handle targets "reconciliation" request (register intend-to-be-registered targets and deregister intend-to-be-registered targets)
+	if err := targetsSynthesizer.Synthesize(ctx); err != nil {
```
Comment (liwenwu-amazon) [inline pkg/deploy/stack_deployer.go:108]: I think it maybe  clear if you
if /* create path */ {
     t.SynthesizeTriggeredTargetGroupsCreation(...)
     create service(...)
}

File: pkg/gateway/model_build_lattice_service.go (hunk 1 above)
Comment (zijun726911) [reply pkg/gateway/model_build_lattice_service.go:85]: Will do

File: pkg/deploy/lattice/target_group_synthesizer_test.go [hunk 10]
```diff
@@ -277,7 +281,13 @@ func Test_SynthersizeTriggeredByServiceImport(t *testing.T) {
 			}
 		}
 		synthesizer := NewTargetGroupSynthesizer(nil, nil, mockTGManager, stack, ds)
-		err := synthesizer.SynthesizeTriggeredTargetGroup(ctx)
+		var err error
+		if !tt.isDeleted {
+			err = synthesizer.SynthesizeTriggeredCreateTargetGroups(ctx)
```
[the remark this answers is not in the document]
Comment (zijun726911) [reply pkg/deploy/lattice/target_group_synthesizer_test.go]: Go it, will change

File: pkg/deploy/lattice/service_synthesizer_test.go [hunk 11]
```diff
@@ -162,8 +162,12 @@ func Test_SynthesizeService(t *testing.T) {
 		}
 
 		synthesizer := NewServiceSynthesizer(mockSvcManager, stack, ds)
-
-		err := synthesizer.Synthesize(ctx)
+		var err error
+		if tt.httpRoute.DeletionTimestamp.IsZero() {
+			err = synthesizer.Synthesize(ctx)
+		} else {
+			err = synthesizer.PostSynthesize(ctx)
```
[the remark this answers is not in the document]
Comment (zijun726911) [reply pkg/deploy/lattice/service_synthesizer_test.go]: Changed, already split them in a the new revision

File: test/suites/integration/httproute_header_match_test.go [hunk 12]
```diff
@@ -15,9 +15,9 @@ import (
 	"time"
 )
 
-var _ = Describe("HTTPRoute header matches", func() {
+var _ = Describe("HTTPRoute header matches", Focus, func() {
```
[the remark this answers is not in the document]
Comment (zijun726911) [reply test/suites/integration/httproute_header_match_test.go]: already removed it in the new revision

File: pkg/deploy/lattice/target_group_synthesizer.go [hunk 13]
```diff
@@ -38,38 +38,32 @@ type targetGroupSynthesizer struct {
 }
 
 func (t *targetGroupSynthesizer) Synthesize(ctx context.Context) error {
-	var ret = ""
 
 	glog.V(6).Infof("Start synthesizing TargetGroupss ...\n")
 
-	if err := t.SynthesizeTriggeredTargetGroup(ctx); err != nil {
-		ret = LATTICE_RETRY
+	if err := t.SynthesizeTriggeredTargetGroupsCreation(ctx); err != nil {
```
[the remark this answers is not in the document]
Comment (zijun726911) [reply pkg/deploy/lattice/target_group_synthesizer.go:44]: ok, will do

File: pkg/deploy/lattice/target_group_synthesizer.go [hunk 14]
```diff
@@ -38,38 +38,32 @@ type targetGroupSynthesizer struct {
...
 
 	t.stack.ListResources(&resTargetGroups)
 
-	glog.V(6).Infof("Synthesize TargetGroups ==[%v]\n", resTargetGroups)
+	glog.V(6).Infof("SynthesizeTriggeredTargetGroupsCreation TargetGroups: [%v]\n", resTargetGroups)
 
 	for _, resTargetGroup := range resTargetGroups {
-
+		if resTargetGroup.Spec.IsDeleted {
+			// in SynthesizeTriggeredTargetGroupsCreation(), we only handle TG Creation request
```
[the remark this answers is not in the document]
Comment (zijun726911) [reply pkg/deploy/lattice/target_group_synthesizer.go:64]: will add

File: pkg/deploy/stack_deployer.go (hunk 9 above)
Comment (zijun726911) [reply pkg/deploy/stack_deployer.go:108]: Got it, will change, although within the serviceSynthesizer.Synthesize() logic itself it could skip deletion deletion requests, but you are right, add a if else here make it more clear

File: pkg/deploy/stack_deployer.go [hunk 15]
```diff
@@ -93,14 +93,48 @@ func NewLatticeServiceStackDeploy(cloud aws.Cloud, k8sClient client.Client, latt
...
+		return err
+	}
+
+	// Handle latticeService creation request
+	if err := serviceSynthesizer.Synthesize(ctx); err != nil {
```
[the remark this answers is not in the document]
Comment (zijun726911) [reply pkg/deploy/stack_deployer.go:113]: serviceSynthesizer now only have  Synthesize() and PostSynthesize() method, should I do a rename:

Synthesize -> SynthesizeCreation ?

File: pkg/deploy/stack_deployer.go [hunk 16]
```diff
@@ -93,14 +93,48 @@ func NewLatticeServiceStackDeploy(cloud aws.Cloud, k8sClient client.Client, latt
...
+		return err
+	}
+
+	//Handle latticeService listener's rules "reconciliation" request
+	if err := ruleSynthesizer.Synthesize(ctx); err != nil {
+		return err
+	}
+
+	//Handle latticeService Deletion request
+	if err := serviceSynthesizer.PostSynthesize(ctx); err != nil {
```
[the remark this answers is not in the document]
Comment (zijun726911) [reply pkg/deploy/stack_deployer.go:128]: serviceSynthesizer now only have Synthesize() and PostSynthesize() method, should I do a rename:

PostSynthesize -> SynthesizeDeletion ?

File: pkg/deploy/stack_deployer.go [hunk 17]
```diff
@@ -93,14 +93,48 @@ func NewLatticeServiceStackDeploy(cloud aws.Cloud, k8sClient client.Client, latt
...
+		return err
+	}
+
+	//Handle targetGroup Deletion request and garbage collection for not-in-use targetGroups
+	if err := targetGroupSynthesizer.PostSynthesize(ctx); err != nil {
```
[the remark this answers is not in the document]
Comment (zijun726911) [reply pkg/deploy/stack_deployer.go:133]: Will split it to:

SynthesizeTriggeredTargetGroupsDeletion()

and 

SynthesizeSDKTargetGroups()

File: pkg/gateway/model_build_lattice_service.go (hunk 1 above)
Comment (zijun726911) [reply pkg/gateway/model_build_lattice_service.go:85]: I found currently we don't have the unit test for  `(t *latticeServiceModelBuildTask) buildModel()` method, that is a large method, I plan to add this unit test in a seperate PR

File: pkg/deploy/stack_deployer.go (hunk 9 above)
Comment (zijun726911) [reply pkg/deploy/stack_deployer.go:108]: I found it's hard to do that, becuase for example, when hit the `targetGroupSynthesizer.SynthesizeTriggeredTargetGroupsCreation()` it need to handle **one list of ** TargetGroups creation or deletion requests (maybe can be a mix of creation and deletion for httproute mutation scenario?), it need to handle the the creation request and skip the deletion requests.

Similarly, for `targetGroupSynthesizer.SynthesizeTriggeredTargetGroupsDeletion()` it need to handle **ONE list of TargetGroups creation and deletion requests**, it need to handle the the deletion request and skip the deletion requests(`resTargetGroup.Spec.IsDeleted && !resTargetGroup.Spec.Config.IsServiceImport`).

Hard to do a if-else statement in `(d *latticeServiceStackDeployer) Deploy()`, because  `targetGroup.Spec.IsDeleted` flag exists for per targetgroup level

File: pkg/gateway/model_build_lattice_service.go (hunk 1 above)
Comment (liwenwu-amazon) [reply pkg/gateway/model_build_lattice_service.go:85]: can u create a issue for this? thanks

File: pkg/deploy/stack_deployer.go (hunk 9 above)
Comment (liwenwu-amazon) [reply pkg/deploy/stack_deployer.go:109]: I meant you can 
* still keep **one list** which  contains new TGs and delete TGs
* SynthesizeTriggeredTargetGroupsCreation() go through the list and takes care of `new TGs`
* SynthesizeTriggeredTargetGroupsDeletion() go through the list and takes care of `deleted TGs`

Comment (liwenwu-amazon) [reply pkg/deploy/stack_deployer.go:109]: NVM , saw your latest update, the change are fine.

Comment (liwenwu-amazon) [reply pkg/deploy/stack_deployer.go:109]: But, we need a unit test to cover func (d *latticeServiceStackDeployer) Deploy(ctx context.Context, stack core.Stack) error {).   Please create an github issue if you are `not` going to cover it in the PR

File: pkg/deploy/lattice/service_synthesizer.go [hunk 18]
```diff
@@ -32,49 +32,47 @@ func (s *serviceSynthesizer) Synthesize(ctx context.Context) error {
 	// TODO
```
Comment (liwenwu-amazon) [inline pkg/deploy/lattice/service_synthesizer.go:32]: remove this comment

File: pkg/deploy/lattice/service_synthesizer.go [hunk 19]
```diff
@@ -32,49 +32,47 @@ func (s *serviceSynthesizer) Synthesize(ctx context.Context) error {
...
-			err := s.serviceManager.Delete(ctx, resService)
-
-			if err == nil {
-				glog.V(6).Infof("service - Synthesizer: finish deleting service %v\n", *resService)
-				s.latticeDataStore.DelLatticeService(resService.Spec.Name, resService.Spec.Namespace)
-
-				// also delete all listeners of this service
-				listeners, err := s.latticeDataStore.GetAllListeners(resService.Spec.Name, resService.Spec.Namespace)
-
-				glog.V(6).Infof("service synthesize -- need to delete listeners %v, err : %v\n", listeners, err)
```
Comment (liwenwu-amazon) [inline pkg/deploy/lattice/service_synthesizer.go:47]: please keep this debug log, which can be helpful

File: pkg/deploy/lattice/target_group_manager.go [hunk 20]
```diff
@@ -160,13 +160,31 @@ func (s *defaultTargetGroupManager) Delete(ctx context.Context, targetGroup *lat
...
 	listResp, err := vpcLatticeSess.ListTargetsAsList(ctx, &listTargetsInput)
-	glog.V(6).Infof("manager delete,  listResp %v, err: %v \n", listResp, err)
+	glog.V(6).Infof("TG manager delete,  listResp %v, err: %v \n", listResp, err)
 	if err != nil {
 		return err
 	}
 
 	var targets []*vpclattice.Target
 	for _, t := range listResp {
+		if t.Status != nil && *t.Status != vpclattice.TargetStatusUnused {
```
Comment (liwenwu-amazon) [inline pkg/deploy/lattice/target_group_manager.go:170]: Please add unit test to cover these

File: pkg/deploy/lattice/target_group_synthesizer.go [hunk 21]
```diff
@@ -38,38 +38,31 @@ type targetGroupSynthesizer struct {
 }
 
 func (t *targetGroupSynthesizer) Synthesize(ctx context.Context) error {
```
Comment (liwenwu-amazon) [inline pkg/deploy/lattice/target_group_synthesizer.go:40]: Please double check if this is impacting `serviceexport` .   `serviceexport` creation and deletion need this one to create or delete target groups.

File: pkg/deploy/lattice/target_group_synthesizer_test.go [hunk 22]
```diff
@@ -31,7 +31,7 @@ import (
 	latticemodel "github.com/aws/aws-application-networking-k8s/pkg/model/lattice"
 )
 
-func Test_SynthesizeTriggeredServiceExport(t *testing.T) {
```
Comment (liwenwu-amazon) [inline pkg/deploy/lattice/target_group_synthesizer_test.go:34]: What happens to unit test for ServiceExport unit test?

File: pkg/deploy/lattice/target_group_synthesizer_test.go [hunk 23]
```diff
@@ -306,6 +283,66 @@ func Test_SynthersizeTriggeredByServiceImport(t *testing.T) {
 	}
 }
 
+func Test_SynthesizeTriggeredTargetGroupsDeletion_TriggeredByServiceImport(t *testing.T) {
```
Comment (liwenwu-amazon) [inline pkg/deploy/lattice/target_group_synthesizer_test.go:286]: ServiceImport should `never` trigger deletion of target group!  Target group should be created or deleted by "ServiceExport".

File: pkg/deploy/lattice/target_group_synthesizer_test.go [hunk 24]
```diff
@@ -751,3 +785,137 @@ func Test_SynthesizeTriggeredService(t *testing.T) {
 
 	}
 }
+
+func Test_SynthesizeTriggeredTargetGroupsDeletion_TriggeredByK8sService(t *testing.T) {
```
Comment (liwenwu-amazon) [inline pkg/deploy/lattice/target_group_synthesizer_test.go:789]: Today, K8S service should `not` trigger target group deletion.

The target group is creation is triggered by HTTPRoute's backendref.  Then, it should be deleted by HTTPRoute deletion or not reference this target group.

File: pkg/deploy/lattice/service_synthesizer.go (hunk 19 above)
Comment (zijun726911) [reply pkg/deploy/lattice/service_synthesizer.go:47]: These code moved to here: `func (s *serviceSynthesizer) SynthesizeDeletion(ctx context.Context)`, 


And I didn't add this code
```
listeners, err := s.latticeDataStore.GetAllListeners(resService.Spec.Name, resService.Spec.Namespace)
for _, l := range listeners {
					s.latticeDataStore.DelListener(resService.Spec.Name, resService.Spec.Namespace,
						l.Key.Port, l.Key.Protocol)
				}
```
Because I think, when we delete the latticeService, we use this order:

- listenerSynthesizer.Synthesize(ctx)
- ruleSynthesizer.Synthesize(ctx)
- serviceSynthesizer.SynthesizeDeletion(ctx)

So that we don't need to clean Listeners DataStore in serviceSynthesizer.SynthesizeDeletion(), is that correct?

File: pkg/deploy/lattice/service_synthesizer.go (hunk 18 above)
Comment (zijun726911) [reply pkg/deploy/lattice/service_synthesizer.go:32]: Will do

File: pkg/deploy/lattice/target_group_synthesizer.go (hunk 21 above)
Comment (zijun726911) [reply pkg/deploy/lattice/target_group_synthesizer.go:40]: Tested the noral serviceexport creation and deletion use cases, it could create and delete the target group

File: pkg/gateway/model_build_lattice_service.go (hunk 1 above)
Comment (zijun726911) [reply pkg/gateway/model_build_lattice_service.go:85]: Ok

File: pkg/deploy/lattice/target_group_synthesizer_test.go (hunk 24 above)
Comment (zijun726911) [reply pkg/deploy/lattice/target_group_synthesizer_test.go:789]: Let me change the each test case name, I intend to test:
For `name: "k8sService triggered target group deletion, ok case"`
it means `name: "httproute deletion with backendref k8sService triggers target group deletion, ok case"`

Comment (zijun726911) [reply pkg/deploy/lattice/target_group_synthesizer_test.go:789]: > deleted by HTTPRoute deletion: 

targetGroupSynthesizer.SynthesizeTriggeredTargetGroupsDeletion(ctx) will handle it

> not reference this target group.

targetGroupSynthesizer.SynthesizeSDKTargetGroups(ctx) will handle it

File: pkg/deploy/lattice/target_group_synthesizer.go [hunk 25]
```diff
@@ -355,7 +330,59 @@ func (t *targetGroupSynthesizer) isTargetGroupUsedByaHTTPRoute(ctx context.Conte
...
+		}
+	}
+	if returnErr {
+		return errors.New(LATTICE_RETRY)
+	} else {
+		return nil
+	}
+}
+
 func (t *targetGroupSynthesizer) PostSynthesize(ctx context.Context) error {
```
Comment (liwenwu-amazon) [inline pkg/deploy/lattice/target_group_synthesizer.go:368]: Are we still using this?  I thought HTTPRoute's synthesizer already takes care of. both K8S HTTPRoute creation and deletion.     

If this is the case, please revert this change.

Comment (zijun726911) [reply pkg/deploy/lattice/target_group_synthesizer.go:368]: latticeServiceStackDeployer.Deploy() (used by httproute controller) uses the 
SynthesizeTriggeredTargetGroupsDeletion and SynthesizeSDKTargetGroups explcitly

But, latticeTargetGroupStackDeployer.Deploy() (used by  serviceExport controller)  still uses 
targetGroupSynthesizer.PostSynthesize()

File: pkg/deploy/lattice/target_group_manager.go (hunk 20 above)
Comment (zijun726911) [reply pkg/deploy/lattice/target_group_manager.go:170]: Will create a issue and add it in a seperate PR

File: pkg/deploy/lattice/target_group_synthesizer.go (hunk 25 above)
Comment (liwenwu-amazon) [reply pkg/deploy/lattice/target_group_synthesizer.go:368]: There is `no` need to change the `targetgroupSynthesize.PostSynthesize()` internal business logic for serviceexport. and also there is `no` need to change `targetgroupSynthesizer.Synthetize() internal business logic too for serviceexport.
